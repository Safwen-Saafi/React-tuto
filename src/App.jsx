import { useDocumentClick } from "./Utils/Hooks/useDocumentClick";
// import { UserContext } from "./Context/UserContext";
// import { PostContainer } from "./Components/Context API/PostContainer";
// import FetchUser from "./Components/Fetch API/FetchUser";
// import { Link, Outlet } from "react-router-dom";
import { ButtonClaude } from "./Components/Use State/ButtonClaude";

export default function App() {
  // Custom Hook
  useDocumentClick();
  // So if you go to the network section, you'll notice how the first request is aborted because the component is unmounted, remember
  // that we are in strictMode so the re-rendering happens twice

  // Context API data
  const users = [
    {
      id: 1,
      username: "saf123",
      email: "safwen@example.com",
    },
    {
      id: 2,
      username: "bob456",
      email: "bob@example.com",
    },
    {
      id: 3,
      username: "sawsan",
      email: "sawsan@example.com",
    },
  ];

  return (
    <div>
      {/* <nav>
      <ul>
        <li><Link to="/form">Users</Link></li>
      </ul>
      </nav> */}

      {/* <UserContext.Provider value={{ ...userdata, setUserData }}>
      </UserContext.Provider>
      <FetchUser/>
      <Outlet /> */}

      {users.map((person) => {
        return <ButtonClaude key={person.id} user={person} />;
      })}
    </div>
  );
}
