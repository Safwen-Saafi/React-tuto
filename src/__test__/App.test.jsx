import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App from "../App";

describe("Should render Users", () => {
  it("should make a snapshot of the component", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });
  describe("Edit button is clicked", () => {
    it("should render save button", async () => {
      const user = userEvent.setup();
      render(<App />);

      const editButtons = screen.getAllByRole("button", { name: "Edit" });

      // Test each edit button
      for (let i = 0; i < editButtons.length; i++) {
        await user.click(editButtons[i]);
        const saveButton = screen.getAllByRole("button", { name: "Save" })[i];
        expect(saveButton).toBeInTheDocument();
      }
    });

    it("should display username & email input fields", async () => {
      const user = userEvent.setup();
      render(<App />);

      const editButtons = screen.getAllByRole("button", { name: "Edit" });
      console.log(editButtons);
      for (let i = 0; i < editButtons.length; i++) {
        await user.click(editButtons[i]);

        // Check that both username and email input fields appear for each edited user
        const usernameInput = screen.getAllByRole("textbox", {
          name: "username",
        })[i];
        const emailInput = screen.getAllByRole("textbox", { name: "email" })[i];

        expect(usernameInput).toBeInTheDocument();
        expect(emailInput).toBeInTheDocument();
      }
    });
  });
});
