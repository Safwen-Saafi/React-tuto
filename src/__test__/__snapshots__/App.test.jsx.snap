// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Should render Users > should make a snapshot of the component 1`] = `
<div>
  <div>
    <div
      style="border: 1px solid #ddd; margin: 10px; padding: 10px; border-radius: 5px;"
    >
      <div>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
      <div>
        <b>
          ID: 
        </b>
        <span>
          1
        </span>
        <br />
        <b>
          Username: 
        </b>
        <span>
          saf123
        </span>
        <br />
        <b>
          Email: 
        </b>
        <span>
          safwen@example.com
        </span>
        <br />
      </div>
    </div>
    <div
      style="border: 1px solid #ddd; margin: 10px; padding: 10px; border-radius: 5px;"
    >
      <div>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
      <div>
        <b>
          ID: 
        </b>
        <span>
          2
        </span>
        <br />
        <b>
          Username: 
        </b>
        <span>
          bob456
        </span>
        <br />
        <b>
          Email: 
        </b>
        <span>
          bob@example.com
        </span>
        <br />
      </div>
    </div>
    <div
      style="border: 1px solid #ddd; margin: 10px; padding: 10px; border-radius: 5px;"
    >
      <div>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
      <div>
        <b>
          ID: 
        </b>
        <span>
          3
        </span>
        <br />
        <b>
          Username: 
        </b>
        <span>
          sawsan
        </span>
        <br />
        <b>
          Email: 
        </b>
        <span>
          sawsan@example.com
        </span>
        <br />
      </div>
    </div>
  </div>
</div>
`;

exports[`Should render Users > should render users in order 1`] = `
<div>
  <div>
    <div
      style="border: 1px solid #ddd; margin: 10px; padding: 10px; border-radius: 5px;"
    >
      <div>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
      <div>
        <b>
          ID: 
        </b>
        <span>
          1
        </span>
        <br />
        <b>
          Username: 
        </b>
        <span>
          saf123
        </span>
        <br />
        <b>
          Email: 
        </b>
        <span>
          safwen@example.com
        </span>
        <br />
      </div>
    </div>
    <div
      style="border: 1px solid #ddd; margin: 10px; padding: 10px; border-radius: 5px;"
    >
      <div>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
      <div>
        <b>
          ID: 
        </b>
        <span>
          2
        </span>
        <br />
        <b>
          Username: 
        </b>
        <span>
          bob456
        </span>
        <br />
        <b>
          Email: 
        </b>
        <span>
          bob@example.com
        </span>
        <br />
      </div>
    </div>
    <div
      style="border: 1px solid #ddd; margin: 10px; padding: 10px; border-radius: 5px;"
    >
      <div>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
      <div>
        <b>
          ID: 
        </b>
        <span>
          3
        </span>
        <br />
        <b>
          Username: 
        </b>
        <span>
          sawsan
        </span>
        <br />
        <b>
          Email: 
        </b>
        <span>
          sawsan@example.com
        </span>
        <br />
      </div>
    </div>
  </div>
</div>
`;
